worlds/dark_souls_3/__init__.py

```diff
@@ -185,364 +223,1350 @@ def create_connection(from_region: str, to_region: str):
+        """
+        # If Yhorm is at Iudex Gundyr, Storm Ruler must be randomized, so it can always be moved.
+        # Fill this manually so that, if very few slots are available in Cemetery of Ash, this
+        # doesn't get locked out by bad rolls on the next two fills.
+        if self.yhorm_location.name == "Iudex Gundyr":
+            self._fill_local_item("Storm Ruler", ["Cemetery of Ash"],
+                                  lambda location: location.name != "CA: Coiled Sword - boss drop")
+
+        # If the Coiled Sword is vanilla, it is early enough and doesn't need to be placed.
+        # Don't place this in the multiworld because it's necessary almost immediately, and don't
+        # mark it as a blocker for HWL because having a miniscule Sphere 1 screws with progression balancing.
+        if self._is_location_available("CA: Coiled Sword - boss drop"):
+            self._fill_local_item("Coiled Sword", ["Cemetery of Ash", "Firelink Shrine"])
+
+        # If the HWL Raw Gem is vanilla, it is early enough and doesn't need to be removed. If
+        # upgrade smoothing is enabled, make sure one raw gem is available early for SL1 players
+        if (
+            self._is_location_available("HWL: Raw Gem - fort roof, lizard")
+            and self.options.smooth_upgrade_items
+        ):
+            self._fill_local_item("Raw Gem", [
+                "Cemetery of Ash",
+                "Firelink Shrine",
+                "High Wall of Lothric"
+            ])
+
+    def _fill_local_item(
+        self, name: str,
+        regions: List[str],
+        additional_condition: Optional[Callable[[DarkSouls3Location], bool]] = None,
+    ) -> None:
+        """Chooses a valid location for the item with the given name and places it there.
+        
+        This always chooses a local location among the given regions. If additional_condition is
+        passed, only locations meeting that condition will be considered.
+
+        If the item could not be placed, it will be added to starting inventory.
+        """
+        item = next((item for item in self.local_itempool if item.name == name), None)
+        if not item: return
+
+        candidate_locations = [
+            location for location in (
+                self.multiworld.get_location(location.name, self.player)
+                for region in regions
+                for location in location_tables[region]
+                if self._is_location_available(location)
+                and not location.missable
+                and not location.conditional
+                and (not additional_condition or additional_condition(location))
+            )
+            # We can't use location.progress_type here because it's not set
+            # until after `set_rules()` runs.
+            if not location.item and location.name not in self.all_excluded_locations
+            and location.item_rule(item)
+        ]
+
+        self.multiworld.itempool.remove(item)
```

Comment:
This is somewhat slow, is there a reason you can't wait to add self.local_itempool to self.multiworld.itempool until after this function runs, since those happen back to back in create_items anyway? Unless this function gets called somewhere else, but in that case, I'd potentially have a different problem with this